webapp/src/shared/components/Game/Game.tsx

```diff
@@ -1,78 +1,157 @@
 import Text, { TextVariant } from '../Text/Text';
 import GameSpinner from '../GameSpinner/GameSpinner';
 import Header from '../Header/Header';
 import { IconVariant } from '../Icon/Icon';
 import { useNavigation } from '../../hooks/UseNavigation';
 
 import './Game.css';
+import socket from '../../socket';
+import { GameInfo, WsException } from '../../types';
+import { useAuth } from '../../hooks/UseAuth';
+import { useNotificationContext } from '../../context/NotificationContext';
+import { useNavigate } from 'react-router-dom';
+import { PLAY_URL } from '../../urls';
+import Button, { ButtonVariant } from '../Button/Button';
+
+const GAME_COMMAND = 'gameCommand';
+const UPDATE_GAME = 'updateGame';
+const JOIN_GAME = 'joinGame';
+const LEAVE_GAME = 'leaveGame';
+const GAME_JOINED = 'gameJoined';
+const GAME_NOT_FOUND = 'gameNotFound';
+const GAME_FINISHED = 'gameFinished';
 
-const Play = () => {
-  const { renderFrame } = useGameAnimation();
+type GameProps = {
+  gameId: string;
+};
+
+const Play = ({ gameId }: GameProps) => {
+  const { notify, warn } = useNotificationContext();
+  const { renderMultiplayerFrame } = useGameAnimation();
   const canvasRef = React.useRef<HTMLCanvasElement>(null);
-  const requestFrameRef = React.useRef<number | null>(null);
   const [score, setScore] = React.useState<number>(0);
-  const { sendGameCommand, updateGame } = useOnlineGame();
-  // Refactor para decorar o hacer HOC con estos componentes
-  // De esta forma usar versión online de los controles y versión online del gameEngine
-  // usamos hook online dentro de la versión decorada
+  const [opponentScore, setOpponentScore] = React.useState<number>(0);
+  const isPlayerOneRef = React.useRef(true);
+  const isPlayerRef = React.useRef(false);
+  const gameStateRef = React.useRef<GameState | null>(null);
+  const { authUser } = useAuth();
+  const navigate = useNavigate();
+  const [gameJoined, setGameJoined] = React.useState(false);
+  const sendGameCommand = React.useCallback(
+    (command: GameCommand) => {
+      if (isPlayerRef.current && gameJoined) {
+        socket.emit(GAME_COMMAND, {
+          command,
+          gameRoomId: gameId,
+        });
+      }
+    },
+    [gameId, gameJoined],
+  );
   useGameControls(sendGameCommand);
-  const { runGameFrame } = useGameEngine(updateGame);
 
-  const gameLoop = React.useCallback(() => {
-    const canvasContext = canvasRef.current?.getContext('2d');
-    const gameState = runGameFrame();
+  const joinGame = () => {
+    if (!gameJoined) {
+      socket.emit(JOIN_GAME, { gameRoomId: gameId });
+    }
+  };
 
-    setScore(gameState.score);
-    canvasContext && renderFrame(canvasContext, gameState);
-    window.requestAnimationFrame(() => gameLoop());
-  }, [runGameFrame, renderFrame]);
+  React.useEffect(() => {
+    return () => {
+      if (gameJoined) {
+        socket.emit(LEAVE_GAME, { gameRoomId: gameId });
+      }
+    };
+  }, [gameId, gameJoined]);
```

Comment:
I see that this works, because the game kind of works, but I don't understand how.

We enter this `Play` function, and we initialize gameJoin to false with useState. Eventually we click, and the gameJoin becomes true. Then this useEffect should be triggered with gameJoin: true, ... which should have the effect of emitting a message to leave the game!

I see this is working, but I don't understand how.

Reply:
Updated. I've removed this condition and moved the leave game emit in the cleanup function of the other `useEffect`.